docs/functions.rst

```diff
@@ -122,6 +136,85 @@ To rename all functions, set the template in the toplevel *options*::
       declarations:
       - decl: void initialize
 
+C++ allows allows overloaded functions and will mangle the names
+behind the scenes.  With Fortran, the mangling must be explicit. To
+accomplish this Shroud uses the *function_suffix* format string.  By
+default, Shroud will use a sequence number.  By explicitly setting
+*function_suffix*, a more meaningful name can be provided:
+
+.. example from tutorial.yaml
+.. code-block:: yaml
+
+  - decl: void Function6(const std::string& name)
+    format:
+      function_suffix: _from_name
+  - decl: void Function6(int indx)
+    format:
+      function_suffix: _from_index
+
+This will create the Fortran functions ``function6_from_name`` and
+``function6_from_index``.  A generic interface named ``function6``
+will also be created which will include the two generated functions.
+
+Likewise, default arguments will produce several Fortran wrappers and
+a generic interface for a single C++ function. The format dictionary
+only allows for a single *function_default* per function.  Instead the
+field *default_arg_suffix* can be set.  It contains a list of
+*function_suffix* values which will be applied from the minimum to the
+maximum number of arguments:
+
+.. example from tutorial.yaml
+.. code-block:: yaml
+
+  - decl: int overload1(int num,
+            int offset = 0, int stride = 1)
+    default_arg_suffix:
+    - _num
+    - _num_offset
+    - _num_offset_stride
+
+Finally, multiple Fortran wrappers can be generated from a single
+templated function. Each instantiation will generate an additional
+Fortran Wrapper and can be distinguished by the *template_suffix*
+entry of the format dictionary.
+
+If there is a single template argument, then *template_suffix* will be
+set to the *flat_name* field of the instantiated argument.  For
+example, ``<int>`` defaults to ``_int``.  This works well for POD types.
+The entire qualified name is used.  For ``<std::string>`` this would be
+``std_string``.  Classes which are deeply nested can produce very long
+values for *template_suffix*. To deal with this, the
+*function_template* field can be set on Class declarations:
+
+.. code-block:: yaml
+
+    - decl: namespace internal
+      declarations:
+      - decl: class ImplWorker1
+        format:
+          template_suffix: instantiation3
+
+By default ``internal_implworker1`` would be used for the
+*template_suffix*.  But in this case ``instantiation3`` will be used.
```

Comment:
Just to confirm I understand this - if I later use `internal::ImplWorker1` as an instantiation for some other function elsewhere in the interface, it will use the suffix `_instantiation3`?

Reply:
correct.